lib/cdo/i18n_string_url_tracker.rb

```diff
@@ -19,16 +66,83 @@ def log(string_key, url, source)
     url = normalize_url(url)
     return unless string_key && url && source
 
-    # record the string : url association.
-    FirehoseClient.instance.put_record(
-      :i18n,
-      {url: url, string_key: string_key, source: source}
-    )
+    add_to_buffer(string_key, url, source)
+  end
+
+  # Sends the buffered i18n string usage data to Firehose.
+  # This should only be called by this class or a unit test.
+  def flush
+    buffer = nil
+
+    # Capture the current data and start a new buffer
+    @buffer.synchronize do # make sure this is the only thread modifying @buffer
+      buffer = @buffer
+      @buffer = {}
+      @buffer.extend(MonitorMixin) # Adds synchronization
+      @buffer_size = 0
+    end
+
+    # log every <string_key>:<url>:<source> combination to Firehose
+    buffer&.each_key do |url|
+      buffer[url].each_key do |string_key|
+        buffer[url][string_key].each do |source|
+          # record the string : url association.
+          FirehoseClient.instance.put_record(
+            :i18n,
+            {url: url, string_key: string_key, source: source}
+          )
+        end
+      end
+    end
+  end
+
+  # Clear the buffer and stop the periodic upload of i18n usage data.
+  # This should only be used by unit tests.
```

Comment:
If these methods should only be used by unit tests, is it possible to put them in a test class/file?

Reply:
I ended up making these methods private instead. I couldn't think of a clean way to make the helper methods live in the test files since they modify the internal state of `I18nStringUrlTracker`